### ISSUE TITLE
UI layout - threadmark button & next previous.

### ISSUE THREAD

**Issue by Xon** [Sidane/xenforo-threadmarks#18]
Hi,

So far I've rolled this threadmark addon to http://forums.sufficientvelocity.com and it has had a great reception.

I'm considering extending adding a Next and Previous button on each post with a threadmark.

From here; http://forums.sufficientvelocity.com/posts/2726004/ one of my users has mentioned the closeness of the threadmark button and the 'go to first unread' button makes it easy to confuse on a mobile device or when operated by memory.

Do you have any thoughts about where a next/previous button could live? And how to move the threadmark button itself so it is more mobile friendly?

The actual code-backend of the next/previous button is underway, and I'm just tidying it up a bunch before it will be ready for a PR.

**Comment by sidane** [comment]
Glad it has been well received on your site. Some very useful feedback and bug reports in that thread.

## Threadmarks button position

I don't think it should move. Threadmarks are part of the thread navigation and belong with the pagination links. This may sound a little harsh, but if somebody's fingers are too fat to tap that button, how do they accurately tap the individual page links which are much smaller?

One option is to the put the threadmarks menu _after_ the `Go to First Unread` button. That way the unread button's position is consistent and user's are less likely to confuse it with the threadmarks button.

## Next/previous button

Nice idea. Positioning the links somewhere near the bottom of the threadmarked post would be best. Once a user finishes reading the post, having the next/previous links nearby would be most convenient.

Next to the `Like` and `Reply` links perhaps?

![Next/previous](http://f.cl.ly/items/0u2X2h0t3q2h0H2V0L0a/next-previous-position.png)

**Comment by Xon** [comment]
# Threadmarks button position

I actually found the threadmark button fairly hard to use on a mobile, as it is nearly exactly the same shape as the "got to unread" and in the same position. That either buttons can pop-up or not is confusing, and moving it to another line may be the best solution.

Going forward it would need to move to support multiple categories. 

IMO, this can be ignored for the short term.

# Next/Previous Button

Yes, I was thinking of the next/previous buttons near the like/reply buttons. But making sure it works on a small screen also requires some testing.

The part of the post gets really cluttered if you have moderator rights, and have a warn/report/edit/delete button.

**Comment by Xon** [comment]
My current thoughts on the next/previous button is; 
- A previous button appears if no more thread marks are in the thread.
- A next button appears if there are more thread marks in the thread.

To allow loading of only the thread marks which land on the page, this implies the thread knows the first and last threadmarks (much like it does post).

This means the UI can just add the buttons depending on if the threadmark post_id matches the first/last of the thread. The link would be to an action for that post (next/previous threadmark), this allows the redirection logic to stay in XF and means each threadmark doesn't need to explicitly know it's neighbour for each marked post.

The actual redirect to the correct ThreadMark would occur in actionNextThreadmark()/actionPreviousThreadMark() actions.

Doing some prototyping on the queries, I've noticed some poor behaviour when linking from threadmarsk -> xf_post to sorting based on thread_id & position. I was seeing the number of rows being touched being the same as the number of posts between threadmarks.

With a couple threads hitting +30000 posts on SV, and more on http://forums.spacebattles.com this performance characteristic needs to be avoided. (I'm also planning on rolling this addon out to SpaceBattles)

As such, moving thread mark positional information into the threadmark table means most of the threadmark queries do not need to touch the xf_post table.

So in summary;
- 2 new fields on the xf_thread table (first+last threadmark)
- 1 new field on threadmarks (position)
- 1 index on thread_id + position on threadmarks.

This will allow:
- per-page loading of threadmarks without needing to fetch the first & last threadmark
- threadmarks do not need to keep pointers to each other, or the entire set load into an array
- Cost of finding the next/previous threadmark is only paid when asking for it.
- Avoids touching the xf_post table to scan for the next post.

**Comment by ottosax** [comment]
A optionaly "Next/previous button" will be a nice feature to let users less scroll up and down and faster go (and find) the next relevant position in a thread.

I will love to see this feature in future in your add-on. :) 

But - please take also a look at board performance... ;)
